src/frontmatter/frontmatter.ts

```diff
@@ -9,24 +9,41 @@ import { IFrontmatter, PropertyPath } from './types';
 import { countNewLines, shiftLines } from './utils';
 
 export class Frontmatter<T extends object = any> implements IFrontmatter<T> {
-  private readonly node: Unist.Literal | null;
-  private readonly root: Unist.Parent;
-  private readonly properties: T | null;
+  public readonly document: Unist.Parent;
+  private readonly node: Unist.Literal;
+  private properties: Partial<T> | null;
 
-  constructor(data: Unist.Parent | string) {
-    const root = typeof data === 'string' ? parseWithPointers(data).data : data;
+  constructor(data: Unist.Parent | string, dirty = true) {
+    const root =
+      typeof data === 'string' ? parseWithPointers(data).data : dirty ? data : JSON.parse(JSON.stringify(data));
```

Comment:
You could avoid deep cloning the entire tree by using something like immer?

Reply:
I hesitated to use Immer as we touch almost every node.
Whenever a change is made, we update all positional info to keep lines in order to make sure diagnostics work well.
```
if (diff !== 0) {
  shiftLines(this.root, diff);
 }
```

Reply:
Note to self: should add a test for new/removed block. I suppose we will need to take 2 extra lines into account and we don't now.

Reply:
Fair enough - this also isn't a hot path (updating frontmatter happens pretty infrequently) so performance isn't really an issue. 👍

Reply:
Yeah, in general, I'm thinking whether we actually do need to shift these lines given the fact we actually send a patch and reparse the content anyway.
Shifting lines is potentially a nice addition, but we don't seem to need it at all and it adds complexity.
What's your take on that?

Reply:
Agree! I think we can ditch it and simplify.